machine-runner/src/runner/runner.ts

```diff
@@ -647,21 +661,72 @@ namespace MachineRunnerIterableIterator {
     MachineName extends string,
     StateUnion extends unknown,
   >({
+    emitter,
     nextValueAwaiter,
     destruction,
+    internals,
   }: {
+    emitter: MachineEmitter<SwarmProtocolName, MachineName, StateUnion>
+    internals: RunnerInternals.Any
     nextValueAwaiter: NextValueAwaiter<
       StateOpaque<SwarmProtocolName, MachineName, string, StateUnion>
     >
     destruction: Destruction
   }): MachineRunnerIterableIterator<SwarmProtocolName, MachineName, StateUnion> => {
-    const onThrowOrReturn = async (): Promise<
-      IteratorResult<StateOpaque<SwarmProtocolName, MachineName, string, StateUnion>, null>
-    > => {
+    type SO = StateOpaque<SwarmProtocolName, MachineName, string, StateUnion>
+
+    const internalsCurrentIsOkForActual = () =>
+      CommandGeneratorCriteria.allOk({
+        isCaughtUp: () => internals.caughtUp && internals.caughtUpFirstTime,
+        isQueueEmpty: () => internals.queue.length === 0,
+        isNotExpired: () => !ImplStateOpaque.isExpired(internals, internals.current),
+        isNotLocked: () => !ImplStateOpaque.isCommandLocked(internals),
+        isNotDestroyed: () => !ImplStateOpaque.isRunnerDestroyed(internals),
+      })
+
+    const onThrowOrReturn = async (): Promise<IteratorResult<SO, null>> => {
       destruction.destroy()
       return Promise.resolve({ done: true, value: null })
     }
 
+    const actual = (): Promise<IteratorResult<SO, null>> => {
+      if (destruction.isDestroyed()) {
+        return Promise.resolve({ done: true, value: null })
+      }
+
+      if (internalsCurrentIsOkForActual()) {
+        return Promise.resolve({
+          done: false,
+          value: ImplStateOpaque.make(internals, internals.current),
+        })
+      }
+
+      const peekOrUnlockPromise = ((): Promise<IteratorResult<SO, null>> => {
+        let unlock = (
+          _: IteratorResult<SO, null>,
+          // eslint-disable-next-line @typescript-eslint/no-empty-function
+        ) => {}
+        const unlockPromise = new Promise<IteratorResult<SO, null>>((res) => (unlock = res))
+        const unlockOnChangeListener = (state: SO) => {
+          if (!internalsCurrentIsOkForActual()) return
+          unlock({ done: false, value: state })
+        }
+        const onDestroy = () => unlock({ done: true, value: null })
+        emitter.on('change', unlockOnChangeListener)
+        destruction.addDestroyHook(onDestroy)
```

Comment:
this should also be undone in the `finally` below

question: why not use `emitter.on('destroyed', ...)`?

Reply:
Thanks for catching this. It completely missed me that this is a potential "memory leak".

Reply:
On second thought, it is on `destruction` and not `emitter.on('destroyed')` because `destruction` lives until the `MachineRunnerIterableIterator` dies. So, on a `noAutoDestroy` clone, `actual` will return `{ done: true }` if it dies regardless of the liveness of the base machine runner